### PR TITLE
fix: PR#5247 by applying review

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #5247

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -1656,11 +1656,18 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                                           size_t &value_index) {
         ASR::ImpliedDoLoop_t *implied_do_loop = ASR::down_cast<ASR::ImpliedDoLoop_t>(implied_do_loop_expr);
 
-        ASR::expr_t* loop_start_expr = ASRUtils::expr_value(implied_do_loop->m_start);
-        ASR::expr_t* loop_end_expr = ASRUtils::expr_value(implied_do_loop->m_end);
+        ASR::expr_t* loop_start_expr = implied_do_loop->m_start;
```
Comment (gxyd) [inline src/lfortran/semantics/ast_common_visitor.h:1659]: I'm sorry, but I still don't understand the fundamental problem that this PR fixes with #5247 .

I see this PR only adding semantic error when a non-parameter start/end/increment is passed for an ImpliedDoLoop in a data statement. Cause, even I wasn't replacing start/end/increment with `expr_value` in-place of the `ImpliedDoLoop` either.

Could you give me something concrete that this PR *helps fix*, which the other PR was doing incorrectly? (let's say: doing a fortran dump or something else maybe?)

Comment (gxyd) [reply src/lfortran/semantics/ast_common_visitor.h:1659]: The difference I see is, mine was:
```cpp
ASR::expr_t* loop_start_expr = ASRUtils::expr_value(implied_do_loop->m_start);
```

the change here is:
```cpp
ASR::expr_t* loop_start_expr = implied_do_loop->m_start;
ASR::expr_t* start_expr_value = ASRUtils::expr_value(loop_start_expr);
```
it seems like, you are doing the same thing that I was, but just in two steps.

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:1659]: I agree and I know this / that PR was not affecting the node in any way unlike what we discussed in PR. So, this PR is just intended to isolate two things and throw error messages :)

**Review by certik — APPROVED** [review]
Seems fine to merge. More work is needed to keep the original expression in ASR, which we can do later.

**Comment by Pranavchiku** [comment]
> More work is needed to keep the original expression in ASR, which we can do later.

Right now we just skip that and add bunch of assignments as shown below:

```console
% cat b.f90 
program data_13
    implicit none
    integer, parameter :: k = 3
    integer :: i, iarx(k), iary(2*k)
    data(iarx(i), i=1, k) / 1, 2, 3 /
end program data_13
% lfortran b.f90 --show-fortran
program data_13
implicit none
integer(4) :: i
integer(4), dimension(3) :: iarx
integer(4), dimension(6) :: iary
integer(4), parameter :: k = 3
iarx(1) = 1
iarx(2) = 2
iarx(3) = 3
end program data_13
```

So here doing `expr_value(loop_start_expr)` won't hurt us. Yes, we need to find a way to unroll this after generating ASR so that we can generate correct code in ASR to Fortran.

**Pull request merged**